### ISSUE TITLE
startRowIndex and maxRows ignored in closure for WorkbookParser.grid()

### ISSUE THREAD

**Issue by marcpa00** [kktec/gsheets#19]
Calling 

```
parser.grid {
   startRowIndex = 1
   maxRows = 1
   columns name: 'string', date:'date', count: 'int', value:'decimal', active: 'boolean'
}
```

on a WorkbookParser created on a workbook visiting the `parsing_demo.xlsx` at top of gsheets sources, will give the 3 lines of sheet instead of only the line with name 'a'.

Strangely enough, all the spock tests where this is tested are in fact green !

I suspect that the closure strategy is maybe modified when running by spock or the `this`, `owner` or `delegate` is different than when run outside of spock...

What I have observed is that if I change the `resolveStrategy` of the closure passed to `grid` to be `Closure.DELEGATE_FIRST`, then I get the expected result, while if I set it to `Closure.OWNER_FIRST`, the default value for new Closures, then changes to startRowIndex or maxRows are ignored.

Forcing resolveStrategy to Closure.DELETAGE_FIRST before calling `closure.call()` in WorkbookParser.grid() also behaves correctly, both in spock tests and in normal context.

I can submit a pull request if you want.

**Comment by kktec** [comment]
I will take a look at this soon. Thanks for bringing it to my attention.

**Comment by marcpa00** [comment]
First, this problems is when gsheets is used in a groovy script.

I think I know what is happening : when WorkbookParser.grid() is called from a script, the owner of the closure parameter is the script.  Therefore, the `startRowIndex = 1` inside the closure ends up being a `setProperty('startRowIndex', 1)` which in turn ends up being a `binding.setVariable('startRowIndex', 1)` on the `binding` object of the script instance.  Therefore, the owner of the closure _receives_ the `startRowIndex = 1`, the delegate, the instance of WorkbookParser where the grid() method is defined, does not.  So, when the `data()` method of the instance of WorkbookParser is called, its property `startRowIndex` is still at 0 because it wasn't modified by the closure passed in parameter to `grid`.

When `grid()` is called from spock tests, the owner is an instance of `org.gsheets.parsing.XmlWorkbookParserSpec` and since this instance has no field `startRowIndex` (and, contrary to a Script, variables are not auto-created on a `setProperty()` invocation), the lookup continues to the delegate which is the instance of `org.gsheets.parsing.WorkbookParser` where the `grid()` and `data()` methods are defined : the property `startRowIndex` of that instance is then modified as expected and the `data()` method sees the modified value.

In other words, the problem is when the owner of closure has a property `startRowIndex` that shadows the one in the delegate (an instance of `WorkbookParser`).  If you add a field `def startRowIndex = 0` to `WorkbookParserSpec`, you will have failed tests.  Forcing the `resolveStrategy` to delegate first (`closure.resolveStrategy = Closure.DELEGATE_FIRST`) before calling `closure.call()` in the `grid()` method will fix this.

**Comment by kktec** [comment]
Interesting, I had not thought of this problem in a scripting context. One approach would be to provide methods in the library for setting variables like this, i.e. `startRowIndex 1`, instead of relying on direct variable access. It would certainly be more DSLish and probably a better technique than modifying variables directly. I presume that setting the delegate variable explicitly in the script closure, i.e. `delegate.startRowIndex = 1`, is a workaround. I will need to think this through a bit when I have more time, hopefully this weekend. Any more suggestions or observations are welcome.

**Comment by marcpa00** [comment]
Yes `delegate.startRowIndex = 1` is a workaround, I tried that and it worked.

I think setting `closure.resolveStrategy = Closure.DELEGATE_FIRST` in `grid()` makes sense as changing `startRowIndex` is something relevant to parsing one workbook, not necessarly something you want to do accross the board (like a method in the library would suggest).

**Comment by kktec** [comment]
I'm planning on probably pursuing both approaches as I think method accessors are also preferable to direct property access. Another workaround would be to create the the Closure and set its 'resolveStrategy' directly before passing it to the grid method. *Do you have any preference for a particular version ?* If not I will probably apply it to 0.4.2 and release that.

**Comment by marcpa00** [comment]
0.4.2 is fine with me !  Any issue preventing bumping groovy to 2.4.5 ?

**Comment by kktec** [comment]
I will probably still compile with Groovy 2.3 as I still want to be sure I can use it with Grails 2.4.x. You should be able to use it with Groovy 2.4.5 without problems.

**Comment by marcpa00** [comment]
Cool, no problem, thanks for you work !

**Comment by kktec** [comment]
Setting the closure delegate first strategy fixes the problem as you described. Thanks Marc!
